src/parser/Parser/Parser.hs

```diff
@@ -1443,18 +1443,20 @@ expr = notFollowedBy nl >>
         cond <- expression
         reserved "then"
         return $ \thn -> Unless{emeta, cond, thn}
-
-      for = blockedConstruct $ do
+      -- JOY for-comprehension
+      for = blockedConstruct $  do
         emeta <- buildMeta
         reserved "for"
-        name <- Name <$> identifier
-        reservedOp "<-"
-        src <- expression
-        stepMeta <- buildMeta
-        step <- option (IntLiteral stepMeta 1)
-                       (do {reserved "by"; expression})
+        sources <- option [] $ (commaSep getForSource)
         reserved "do"
-        return $ \body -> For{emeta, name, src, step, body}
+        return $ \body -> For{emeta, sources, body}
+        where
+          getForSource = do
+             forVar <- Name <$> identifier
+             let forVarType = Nothing
```

Comment:
you can remove this line and add directly the `Nothing` in the `ForSource` a few lines below.